feat: stable first version (#1)
elastic/static-fs: README.md

[![NPM Version][npm-version-image]][npm-version-url]
[![NODE Version][node-version-image]][node-version-url]
[![David Node Deps Manager][david-node-deps-manager-image]][david-node-deps-manager-url]
[![David Node Deps Manager Development][david-node-deps-manager-dev-image]][david-node-deps-manager-dev-url]
[![NPM Install Size][npm-install-size-image]][npm-install-size-url]

A static filesystem to bundle multiple files into one that are
able to be read by Node.js through `require` or the [fs module](https://nodejs.org/api/fs.html).

## Why

There are a lot of use cases which require shipping `node_modules` along with the distribution in order to achieve a `zero install` workflow for our end users. Given the nature of node_modules, this directory ends up containing a magnitude of files resulting in a degraded experience, especially on Windows environments (bad performance unzipping, bad performance through an installer,  max file path length, etc).

That was the first motivation and the main use case for the static filesystem: 
allow to bundle all the files from the `node_modules` during the build process into a single file 

> and then, in runtime, force node to first look into that statically generated 

Reviewer: ```suggestion
and then, at runtime, force node to first look into that statically generated 
```

Author: 👍

filesystem when searching for a file and only look in the real filesystem 
in case the file is not found on the static one.

While the main use case is the above there could be others like for example shipping almost 
every product file in a statically generated filesystem so we can make the product structure 

> simple and transparent for the end user. The only thing to keep in mind is `that 

Reviewer: Did you mean to make this an inline code-block, or just put the sentence in quotes?

Author: I have used the inline code-block to highlight important parts in the readme

tool was designed to be used during a build process` under certain assumptions.

## Features

- Pack multiple files into a single one called static filesystem that would 
mount relatively to the parent directory where it is generated

- Patches `require`, `fs`, `process` and `child_process` to be able to read 
from the static filesystem

- Run multiple static filesystem per application

- One single build step tool that works out of the box

## Getting Started

Remember, that this is a development tool intended to be used during your 
build process, so install it as `development dependency` and run it as 
a final step once your build produce the raw build artifacts.

### Install

`npm install --save-dev static-fs`

### Usage

**Generate a static filesystem volume**
```javascript
// example_build_step.js

const { generateStaticFsVolume } = require('static-fs');
const { resolve } = require('path');

// one example of a dependency to delete files in bulk, you can use any other
const del = require('del');

(async () => {

>   const mountRoot = resolve('../../'), 

Reviewer: It's unclear from this example why you need to go out two directories. I assume this is because you lifted this from Kibana, where this file exists inside a sub-sub-directory, but from the readers' point of view I think it's just adding confusion.

Author: 👍

    folderToAdd = resolve('../../node_modules'),
    entryPoint = resolve('../../index.js');
  
  // Generate a static filesystem volume
  const addedFiles = await generateStaticFsVolume(
    mountRoot,
    [
      folderToAdd
    ],
    [
      entryPoint
    ]
    //, [] -> exclusions are optional and equal to [] by default
  );
  
  // Delete all the files bundled into the static filesystem volume
  await del(addedFiles, { force: true });
})()
```

**Run your app**

Just run `node index.js` (which was the entryPoint we assume for that example) 
in the root of your distributable app folder and everything should work. 

## API

### `async generateStaticFsVolume(mountRootDir, foldersToAdd, appEntryPointsToPatch, exclusions)`

An async function that would take care of generating the static filesystem 
considering that the root path to mount it would be `mountRootDir`, the content 
would be created according `foldersToAdd`, your application entry points 
would be automatically patched according `appEntryPointsToPatch` so node can read 
from the generated static filesystem, and any path (folder or file) listed on `exclusions`
would be discarded.

> > NOTE: After running that function a folder called `static_fs` would be 

Reviewer: Did you consider naming the folder `.static_fs` instead of `static_fs`?

Author: I thought about that but I think it's better to keep the folder visible don't u think so?

Author: I've decided to go with `.static_fs` after our offline discussion

be a folder or a single file and they should be absolute resolved against 
`mountRootDir` otherwise the function would throw an error. In case the path 
is a folder, the entire folder and children would be also excluded. 

**Returns**

An `string[]` containing the paths of each file and base folder 
that was added into this static filesystem instance. 

## Known Limitations

A little set of things are not supported by the static filesystem. They can 
be summarised in the following list:

- `.node` native modules files are not supported inside the static filesystem. 
They will be excluded during the generation phase and are expected to be present 
in the original location on the real filesystem.

- Like the above mentioned `.node` files, every other platform-specific files that
requires to be compiled, processed or any other changing operation during runtime are 
not supported inside the static filesystem. As they won't be excluded automatically, 
they should not be included inside the static filesystem during the generation phase. 

- While this is more a design choice than a limitation we choose to list it 

> here: any write operation is not supported against the static filesystem during runtime.

Reviewer: Consider mentioning briefly what will happen if you attempt to write to one of those files

Author: done 👍

It would mimic the file state at the time it was generated.

Let us know if you found other unknown limitations [opening an issue](https://github.com/elastic/static-fs/issues/new).

## Contributing

If you would like to contribute, please read [CONTRIBUTING](https://github.com/elastic/static-fs/blob/master/CONTRIBUTING.md).

## License

See [LICENSE](https://github.com/elastic/static-fs/blob/master/LICENSE).

## FAQ 

Read the [FAQ](https://github.com/elastic/static-fs/blob/master/FAQ.md) for details.

## Thanks

> Our thanks can be read at [THANKS](https://github.com/elastic/static-fs/blob/master/THANKS.md).

Reviewer: Tip: You don't have to write out the full URL when linking to files or folders in the same repo/branch on GitHub. You can just use a relative path:

```suggestion
Our thanks can be read at [THANKS](THANKS.md).
```

This means that it will also work on forks of the project _and_ in other branches on the same project :)

Author: 👍

<h1 align="center">
  static-fs
</h1> 

[![CI Build Status][ci-build-status-image]][ci-build-status-url]
[![Commitizen Friendly][commitizen-friendly-image]][commitizen-friendly-url]
[![NPM Version][npm-version-image]][npm-version-url]
[![NODE Version][node-version-image]][node-version-url]
[![David Node Deps Manager][david-node-deps-manager-image]][david-node-deps-manager-url]
[![David Node Deps Manager Development][david-node-deps-manager-dev-image]][david-node-deps-manager-dev-url]
[![NPM Install Size][npm-install-size-image]][npm-install-size-url]

> A static filesystem to bundle multiple files into one that are lately

Reviewer 2: ```suggestion
A static filesystem to bundle multiple files into one that are
```

able to be read by Node.js through `require` or the [fs module](https://nodejs.org/api/fs.html).

## Why

There are a lot of use cases when we have the need to ship the `node_modules` 
along with the distribution files of our code in order to achieve a `zero install` 
workflow for our end users. As the application evolves so the `node_modules` number 
of files which would cause a lot of problems specially on `windows` environments 

> (bad performance unzipping, bad performance through an installer,  max file path length, etc).

Reviewer 2: ```suggestion
There are a lot of use cases which require shipping `node_modules` along with the distribution in order to achieve a `zero install` workflow for our end users. Given the nature of node_modules, this directory ends up containing a magnitude of files resulting in a degraded experience, especially on Windows environments (bad performance unzipping, bad performance through an installer,  max file path length, etc).
```

allow to bundle all the files from the `node_modules` during the build process into a single file 
and then, in runtime, force node to first look into that statically generated 
filesystem when searching for a file and only look in the real filesystem 
in case the file is not found on the static one.

While the main use case is the above there could be others like for example shipping almost 
every product file in a statically generated filesystem so we can make the product structure 
simple and transparent for the end user. The only thing to keep in mind is `that 
tool was designed to be used during a build process` under certain assumptions.

## Features

- Pack multiple files into a single one called static filesystem that would 
mount relatively to the parent directory where it is generated

- Patches `require`, `fs`, `process` and `child_process` to be able to read 
from the static filesystem

- Run multiple static filesystem per application

- One single build step tool that works out of the box

## Getting Started

Remember, that is a development tool intended to be used during your 
build process, so install it as `development dependency` and run it as 

> a one final step once your build produce the raw distributable artifacts.

Reviewer 2: ```suggestion
Remember, that this is a development tool intended to be used during your 
build process, so install it as `development dependency` and run it as 
a final step once your build produce the raw build artifacts.
```

### Install

> `npm install --save-dev static-fs`

Reviewer: Nit: I personally like having this render as a code block. If this wasn't your intention, feel free to ignore this comment 😄 

````suggestion
```
npm install --save-dev static-fs
```
````

Author: 👍

**Returns**

An `string[]` containing the paths of each file and base folder 
that was added into this static filesystem instance. 

## Known Limitations

A little set of things are not supported by the static filesystem. They can 
be summarised in the following list:

- `.node` native modules files are not supported inside the static filesystem. 
They will be excluded during the generation phase and are expected to be present 
in the original location on the real filesystem.

- Like the above mentioned `.node` files, every other platform-specific files that
requires to be compiled, processed or any other changing operation during runtime are 
not supported inside the static filesystem. As they won't be excluded automatically, 
they should not be included inside the static filesystem during the generation phase. 

- While this is more a design choice than a limitation we choose to list it 
here: any write operation is not supported against the static filesystem during runtime.
It would mimic the file state at the time it was generated.

Let us know if you found other unknown limitations [opening an issue](https://github.com/elastic/static-fs/issues/new).

## Contributing

> If you wanna contribute please read [CONTRIBUTING.md](https://github.com/elastic/static-fs/blob/master/CONTRIBUTING.md).

Reviewer 2: ```suggestion
If you would like to contribute, please read [CONTRIBUTING.md](https://github.com/elastic/static-fs/blob/master/CONTRIBUTING.md).
```
